provider/ultradns_test.go

```diff
@@ -0,0 +1,658 @@
+		}
+		err := provider.ApplyChanges(context.Background(), changes)
+		if err == nil {
+			t.Errorf("We wanted it to fail since multiple AAAA targets are not allowed")
+		}
+		resp, _ := provider.client.Do("GET", "https://api.ultradns.com/zones/kubernetes-ultradns-provider-test.com./rrsets/AAAA/ttl.kubernetes-ultradns-provider-test.com.", nil, udnssdk.RRSetListDTO{})
+		assert.Equal(t, resp.Status, "404 Not Found")
+	}
+}
+
+// Test case to check multiple CNAME targets.
+func TestUltraDNSProvider_MultipleTargetCNAME(t *testing.T) {
+	_, ok := os.LookupEnv("ULTRADNS_INTEGRATION")
+	if !ok {
+		log.Printf("Skipping test")
+
+	} else {
+		provider, err := NewUltraDNSProvider(endpoint.NewDomainFilter([]string{"kubernetes-ultradns-provider-test.com"}), false)
+		changes := &plan.Changes{}
+
+		changes.Create = []*endpoint.Endpoint{
+			{DNSName: "ttl.kubernetes-ultradns-provider-test.com", Targets: endpoint.Targets{"nginx.loadbalancer.com.", "nginx1.loadbalancer.com."}, RecordType: "CNAME", RecordTTL: 100},
+		}
+		err = provider.ApplyChanges(context.Background(), changes)
+		if err == nil {
+			t.Errorf("We wanted it to fail since multiple CNAME targets are not allowed")
+		}
+
+		resp, _ := provider.client.Do("GET", "https://api.ultradns.com/zones/kubernetes-ultradns-provider-test.com./rrsets/CNAME/kubernetes-ultradns-provider-test.com.", nil, udnssdk.RRSetListDTO{})
+		assert.Equal(t, resp.Status, "404 Not Found")
+	}
+}
+
+//Testing creation of RD Pool
+func TestUltraDNSProvider_newRDPoolObjectCreation(t *testing.T) {
+	mocked := mockUltraDNSRecord{nil}
+	mockedDomain := mockUltraDNSZone{nil}
+
+	provider := &UltraDNSProvider{
+		client: udnssdk.Client{
+			RRSets: &mocked,
+			Zone:   &mockedDomain,
+		},
+	}
+	changes := &plan.Changes{}
+	changes.UpdateNew = []*endpoint.Endpoint{{DNSName: "kubernetes-ultradns-provider-test.com.", Targets: endpoint.Targets{"1.1.2.2", "192.168.0.24"}, RecordType: "A", RecordTTL: 100}}
+	changesList := &UltraDNSChanges{
+		Action: "UPDATE",
+		ResourceRecordSetUltraDNS: udnssdk.RRSet{
+			RRType:    "A",
+			OwnerName: "kubernetes-ultradns-provider-test.com.",
+			RData:     []string{"1.1.2.2", "192.168.0.24"},
+			TTL:       100,
+		},
+	}
+	rdPoolObject := udnssdk.RDPoolProfile{
+		Context:     udnssdk.RDPoolSchema,
+		Order:       "ROUND_ROBIN",
+		Description: "kubernetes-ultradns-provider-test.com.",
+	}
+
+	actualRDPoolObject, _ := provider.newRDPoolObjectCreation(context.Background(), changesList)
+	assert.Equal(t, rdPoolObject, actualRDPoolObject)
+
+}
+
+//Testing Failure scenarios over NewUltraDNS Provider
+func TestNewUltraDNSProvider_FailCases(t *testing.T) {
+	_ = os.Setenv("ULTRADNS_USERNAME", "")
+	_ = os.Setenv("ULTRADNS_PASSWORD", "")
+	_ = os.Setenv("ULTRADNS_BASEURL", "")
+	_ = os.Setenv("ULTRADNS_ACCOUNTNAME", "")
+	_ = os.Setenv("ULTRADNS_POOL_TYPE", "xyz")
+	_, err := NewUltraDNSProvider(endpoint.NewDomainFilter([]string{"test-ultradns-provider.com"}), true)
+	if err == nil {
```

Comment:
use assert.Nil or assert.NotNil instead.

Reply:
yup used Assert.Nil and NotNil every where